### ISSUE TITLE
Custom LogitsProcessor/LogitsWarper

### ISSUE THREAD

**Issue by loganlebanoff** [huggingface/text-generation-inference#1050]
### Feature request

I have my own implementation of a LogitsWarper in this format https://huggingface.co/docs/transformers/internal/generation_utils#transformers.LogitsProcessor that I would like to add to TGI. There are already warpers for temperature, top_p, etc., but I want to add my own. Note: I don't want my LogitsWarper to be merged in to the repo, since I have a vey specific use case -- I just want to know how it can be done.

### Motivation

I want to have the benefits of TGI like batching and tensor parallelism while also adding custom LogitsWarper functionality.

### Your contribution

I've tried editing the source files at `text-generation-inference/server/text_generation_server/utils/logits_process.py` etc., but I started going down a rabbit hole of files that I needed to edit.

**Comment by abhinavkulkarni** [comment]
Yes, this will be very much appreciated. This would be very useful to force the model to generate a JSON response.

Something like this would be quite useful if and when TGI decided to add support for grammar based sampling. Recently `llama.cpp` [added support](https://github.com/ggerganov/llama.cpp/blob/master/grammars/README.md) for the same.

**Comment by abhinavkulkarni** [comment]
@Narsil: Some guidance on this would be very much appreciated. Wether this would be a desired feature and if so, where to start in the codebase.

Thanks!

**Comment by mjsteele12** [comment]
@abhinavkulkarni I almost added grammar support as an idea in discussions before I found this thread/your comment. I am extremely interested in merging the [outlines](https://github.com/outlines-dev/outlines) project with TGI. +1 to this idea/functionality!

**Comment by abhinavkulkarni** [comment]
Thanks @mjsteele12, here's similar project: https://github.com/Shopify/torch-grammar

It would be good to get some guidance from TGI maintainers as to how to go about this.

**Comment by abhinavkulkarni** [comment]
[Here's](https://github.com/ggerganov/llama.cpp#constrained-output-with-grammars) how Llama.cpp does it. More BNF grammars from [here](https://github.com/ggerganov/llama.cpp/tree/master/grammars).

**Comment by tripathiarpan20** [comment]
@abhinavkulkarni @mjsteele12 The idea to add custom grammar for generations in TGI is of great interest to me too, any updates on this? any way I can help?

**Comment by loganlebanoff** [comment]
I got a custom LogitsWarper working! Here's the changes I made, I hope it can help anyone else who is working on something similar. 
https://github.com/huggingface/text-generation-inference/compare/main...loganlebanoff:text-generation-inference:main

**Comment by OlivierDehaene** [comment]
Hey!
I'm OOL on this. 
What would be the minimal API change that you guys would need? If we only add `logit_bias`, would that be enough?

**Comment by OlivierDehaene** [comment]
What's the standard? BNF? EBNF? What about Guidance?

**Comment by abhinavkulkarni** [comment]
@OlivierDehaene: I'm light on the details about Guidance, but it should be possible to integrate an EBNF grammar parser and state machine (SM) in TGI. There is a PR currently open on `transformers` repo ([link](https://github.com/huggingface/transformers/pull/27557)) that adds a parser and SM in a custom logits processor.

The user simply needs to provide an EBNF grammar file.

Recently, `llama.cpp` integrated this functionality ([link](https://github.com/ggerganov/llama.cpp/blob/master/grammars/README.md)).

**Comment by lapp0** [comment]
This functionality is critical for function calling, generating code and data serialization formats such as json, yaml, and csv, along with providing other application-specific constraints on the generation.

Some commentary I've found on logit processing:

Logit processing in the context of returning `top_n_tokens` resulted in a performance degradation of 25% per https://github.com/huggingface/text-generation-inference/pull/617#issuecomment-1682469471 

Although my understanding is that in certain cases grammar-constrained sampling would improve speed, as the candidate space is reduced.

Breaking handling logit processing via a series of smaller requests for the model is marked as undesirable by Narsil https://github.com/huggingface/text-generation-inference/pull/810#issuecomment-1681903799

Guidance is marked as a priority for the next release https://github.com/huggingface/text-generation-inference/issues/505#issuecomment-1708367609 however this was at the time of of v1.0.3, so I'm not sure whether this is still considered a priority 3 releases later.

Thanks for your hard work on TGI, would be great to hear some input on how this might be integrated.

**Comment by comaniac** [comment]
Upvote for this. Now OpenAI API allows users specify JSON response format. I believe they achieved this by controlling logit processor to follow JSON grammar (i.e., constraint semantic decoding). This, however, cannot be added to TGI for now due to the lack of custom logic processor.

**Comment by abhinavkulkarni** [comment]
I tried integrating code from this [transformers PR](https://github.com/huggingface/transformers/pull/27557) and was able to successfully generate JSON responses from TGI.

**Comment by puppetm4st3r** [comment]
@abhinavkulkarni is there a tentative date to add the solution for the grammar module?
